drivers/auxdisplay/auxdisplay_jhd1313.c

```diff
@@ -0,0 +1,369 @@
+
+/* LCD Display Commands */
+#define JHD1313_CMD_SCREEN_CLEAR	(1 << 0)
+#define JHD1313_CMD_CURSOR_RETURN	(1 << 1)
+#define JHD1313_CMD_INPUT_SET		(1 << 2)
+#define JHD1313_CMD_DISPLAY_SWITCH	(1 << 3)
+#define JHD1313_CMD_CURSOR_SHIFT	(1 << 4)
+#define JHD1313_CMD_FUNCTION_SET	(1 << 5)
+#define JHD1313_CMD_SET_CGRAM_ADDR	(1 << 6)
+#define JHD1313_CMD_SET_DDRAM_ADDR	(1 << 7)
+
+#define JHD1313_DS_DISPLAY_ON		(1 << 2)
+#define JHD1313_DS_CURSOR_ON		(1 << 1)
+#define JHD1313_DS_BLINK_ON		(1 << 0)
+
+#define JHD1313_LED_REG_R		0x04
+#define JHD1313_LED_REG_G		0x03
+#define JHD1313_LED_REG_B		0x02
+
+struct auxdisplay_jhd1313_data {
+	uint8_t input_set;
+	bool power;
+	bool cursor;
+	bool blinking;
+	uint8_t function;
+	uint8_t backlight;
+};
+
+struct auxdisplay_jhd1313_config {
+	struct auxdisplay_capabilities capabilities;
+	struct i2c_dt_spec bus;
+};
+
+static uint8_t colour_define[][4] = {
+	{ 0,   0,   0   },      /* Off */
+	{ 255, 255, 255 },	/* White */
+	{ 255, 0,   0   },      /* Red */
+	{ 0,   255, 0   },      /* Green */
+	{ 0,   0,   255 },      /* Blue */
+};
+
+static void auxdisplay_jhd1313_reg_set(const struct device *i2c, uint8_t addr, uint8_t data)
+{
+	uint8_t command[2] = { addr, data };
+
+	i2c_write(i2c, command, sizeof(command), JHD1313_BACKLIGHT_ADDR);
+}
+
+static int auxdisplay_jhd1313_print(const struct device *dev, const uint8_t *data, uint16_t size)
+{
+	const struct auxdisplay_jhd1313_config *config = dev->config;
+	uint8_t buf[] = { JHD1313_CMD_SET_CGRAM_ADDR, 0 };
+	int rc = 0;
+	int16_t i;
+
+	for (i = 0; i < size; i++) {
+		buf[1] = data[i];
+		rc = i2c_write_dt(&config->bus, buf, sizeof(buf));
+	}
+
+	return rc;
+}
+
+static int auxdisplay_jhd1313_cursor_position_set(const struct device *dev,
+						  enum auxdisplay_position type, int16_t x,
+						  int16_t y)
+{
+	unsigned char data[2];
+
+	if (type != AUXDISPLAY_POSITION_ABSOLUTE) {
+		return -EINVAL;
+	}
+
+	if (y == 0U) {
+		x |= 0x80;
+	} else {
+		x |= 0xC0;
+	}
+
+	data[0] = JHD1313_CMD_SET_DDRAM_ADDR;
+	data[1] = x;
+
+	return i2c_write_dt(&config->bus, data, 2);
+}
+
+static int auxdisplay_jhd1313_clear(const struct device *dev)
+{
+	int rc;
+	const struct auxdisplay_jhd1313_config *config = dev->config;
+	uint8_t clear[] = { 0, JHD1313_CMD_SCREEN_CLEAR };
+
+	rc = i2c_write_dt(&config->bus, clear, sizeof(clear));
+	LOG_DBG("Clear, delay 20 ms");
+	k_sleep(K_MSEC(20));
+
+	return rc;
+}
+
+static int auxdisplay_jhd1313_update_display_state(
+				const struct auxdisplay_jhd1313_config *config,
+				struct auxdisplay_jhd1313_data *data)
+{
+	int rc;
+	uint8_t buf[] = { 0, JHD1313_CMD_DISPLAY_SWITCH };
+
+	if (data->power) {
+		buf[1] |= JHD1313_DS_DISPLAY_ON;
+	}
+
+	if (data->cursor) {
+		buf[1] |= JHD1313_DS_CURSOR_ON;
+	}
+
+	if (data->blinking) {
+		buf[1] |= JHD1313_DS_BLINK_ON;
+	}
+
+	rc = i2c_write_dt(&config->bus, buf, sizeof(buf));
+
+	LOG_DBG("Set display_state options, delay 5 ms");
+	k_sleep(K_MSEC(5));
```

Comment:
magic sleep

Reply:
Not my driver, I just took what was in tree already and have no way to test it

Reply:
Moved to defines